### PR TITLE
Add ProposalStatus dynamic enum

### PR DESCRIPTION
**Pull request by toddburnside**
Also updates the schema and extracts some common code for the dynamic enums.

### CODE REVIEW

File: model/src/main/scala/lucuma/schemas/enums/ProposalAttachmentType.scala [hunk 1]
```diff
@@ -27,21 +27,13 @@ object ProposalAttachmentType:
...
-      .downField("proposalAttachmentTypeMeta")
-      .as[List[ProposalAttachmentType]] match
-      case Left(err)   => err.printStackTrace; throw err
-      case Right(json) => json
-
   // The givens are apparently (probably) constructed lazily.
   // See https://alexn.org/blog/2022/05/11/implicit-vs-scala-3-given/
   // We want to fail immediately if there is a problem, so we'll reference
   // the enumerated givens here.
   Enumerated[ProposalAttachmentType]
```
Comment (cquiroz) [inline model/src/main/scala/lucuma/schemas/enums/ProposalAttachmentType.scala:34]: Odd

**Review by cquiroz — APPROVED** [review]
LGTM

File: model/src/main/scala/lucuma/schemas/enums/ProposalAttachmentType.scala (hunk 1 above)
Comment (toddburnside) [reply model/src/main/scala/lucuma/schemas/enums/ProposalAttachmentType.scala:34]: Yeah. It's an interesting blog post...

File: model/src/main/scala/lucuma/schemas/enums/DynamicEnums.scala [hunk 2]
```diff
@@ -5,9 +5,19 @@ package lucuma.schemas.enums
...
       case Left(err)   => err.printStackTrace; throw err
       case Right(json) => json.hcursor
+
+  def enumeratedInstance[E](
+    field: String,
+    tagFn: E => String
+  )(using Decoder[E]): Enumerated[E] =
+    val values: List[E] = parsedEnums.downField(field).as[List[E]] match
+      case Left(err)   => err.printStackTrace; throw err
+      case Right(list) => list
```
Comment (rpiaggio) [inline model/src/main/scala/lucuma/schemas/enums/DynamicEnums.scala:22]: Nitpicking: you can omit constructing the `values` and return the `Enumerated` instance directly from the `Right` case.

File: model/src/main/scala/lucuma/schemas/enums/ProposalStatus.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  val name: Lens[ProposalStatus, String] = Focus[ProposalStatus](_.name)
+
+  lazy val NotSubmitted: ProposalStatus = Enumerated[ProposalStatus].unsafeFromTag("NOT_SUBMITTED")
+  lazy val Submitted: ProposalStatus    = Enumerated[ProposalStatus].unsafeFromTag("SUBMITTED")
+
+  // The givens are apparently (probably) constructed lazily.
+  // See https://alexn.org/blog/2022/05/11/implicit-vs-scala-3-given/
+  // We want to fail immediately if there is a problem, so we'll reference
+  // the enumerated givens here.
+  Enumerated[ProposalStatus]
```
Comment (rpiaggio) [inline model/src/main/scala/lucuma/schemas/enums/ProposalStatus.scala:30]: Maybe we can reference all the `Enumerated`s from a single file? (And have the lengthy comment once)

**Review by rpiaggio — APPROVED** [review]
LGTM

File: model/src/main/scala/lucuma/schemas/enums/DynamicEnums.scala (hunk 2 above)
Comment (toddburnside) [reply model/src/main/scala/lucuma/schemas/enums/DynamicEnums.scala:22]: Good point

File: model/src/main/scala/lucuma/schemas/enums/ProposalStatus.scala (hunk 3 above)
Comment (toddburnside) [reply model/src/main/scala/lucuma/schemas/enums/ProposalStatus.scala:30]: I'll put them in DynamicEnums

**Pull request merged**